src/hypothesis/searchstrategy/strategies.py

```diff
@@ -203,35 +236,70 @@ class OneOfStrategy(SearchStrategy):
     def __init__(self, strategies, bias=None):
         SearchStrategy.__init__(self)
         strategies = tuple(strategies)
-        self.element_strategies = list(strategies)
+        self.original_strategies = list(strategies)
+        self.__element_strategies = None
         self.bias = bias
+        self.__in_branches = False
         if bias is not None:
             assert 0 < bias < 1
             self.sampler = cu.Sampler(
                 [bias ** i for i in range(len(strategies))])
         else:
             self.sampler = None
 
+    def calc_is_empty(self):
+        return len(self.element_strategies) == 0
+
+    @property
+    def element_strategies(self):
+        from hypothesis.strategies import check_strategy
+        if self.__element_strategies is None:
+            strategies = []
+            for arg in self.original_strategies:
+                check_strategy(arg)
+                if not arg.is_empty:
+                    strategies.extend(
+                        [s for s in arg.branches if not s.is_empty])
+            pruned = []
+            seen = set()
+            for s in strategies:
+                if s is self:
+                    continue
+                if s in seen:
+                    continue
+                seen.add(s)
+                pruned.append(s)
+            self.__element_strategies = pruned
+        return self.__element_strategies
+
     def do_draw(self, data):
         n = len(self.element_strategies)
-        if self.sampler is None:
+        if n == 0:
+            data.mark_invalid()
+        elif n == 1:
+            return self.element_strategies[0].do_draw(data)
+        elif self.sampler is None:
             i = cu.integer_range(data, 0, n - 1)
         else:
             i = self.sampler.sample(data)
 
         return data.draw(self.element_strategies[i])
 
     def __repr__(self):
-        return ' | '.join(map(repr, self.element_strategies))
+        return ' | '.join(map(repr, self.original_strategies))
```

Comment:
Yup. It seemed more sensible than having repr have the side effect of forcing the strategy. I guess we could have the repr change values, but it seemed better for it not to (though the deferred strategies do, so maybe it doesn't matter)